### PR TITLE
remove unverified role on /verify

### PR DESCRIPTION
**Pull request by shobhan-sundar-goutam**
Date: 01/10/24

Developer Name: Shobhan Sundar Goutam

---

## Issue Ticket Number
- Closes <https://github.com/Real-Dev-Squad/discord-slash-commands/issues/234>

## Description
This PR fixes the first issue which is removing unverified role from discord when user runs /verify and connects their RDS Discord

### Documentation Updated?

- [ ] Yes
- [X] No

### Under Feature Flag

- [ ] Yes
- [X] No

### Database Changes

- [ ] Yes
- [X] No

### Breaking Changes

- [ ] Yes
- [X] No

### Development Tested?

- [X] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

https://github.com/user-attachments/assets/98d0a3e3-9a34-429a-81d6-3dd79625eb0a

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

image
image

image
image

</details>

### CODE REVIEW

File: test/integration/external-accounts.test.js [hunk 1]
```diff
@@ -448,12 +450,21 @@ describe("External Accounts", function () {
 
   describe("PATCH /external-accounts/link/:token", function () {
```
Comment (shubhdevelop) [inline test/integration/external-accounts.test.js:451]: Integration test doesn't check if the role actually got removed, add that in the success case

Comment (shobhan-sundar-goutam) [reply test/integration/external-accounts.test.js:451]: I've added it. Can you please check it?

File: controllers/external-accounts.js [hunk 2]
```diff
@@ -69,6 +71,8 @@ const linkExternalAccount = async (req, res) => {
       userId
     );
 
+    await removeDiscordRole(req.userData, attributes.discordId, undefined, "unverified");
```
[the remark this answers is not in the document]
Comment (shobhan-sundar-goutam) [reply controllers/external-accounts.js:74]: In case of failure, the catch block of the function where it is called will send the response.

[the remark this answers is not in the document]
Comment (shobhan-sundar-goutam) [reply controllers/external-accounts.js]: Handled the cases that you were asking for. Can you please check it?

File: controllers/external-accounts.js [hunk 3]
```diff
@@ -69,6 +72,17 @@ const linkExternalAccount = async (req, res) => {
...
+      req.userData,
+      attributes.discordId,
+      undefined,
+      DISCORD_ROLES.UNVERIFIED
+    );
+
+    if (!isUnverifiedRoleRemoved) {
+      return res.status(500).json({ message: "Role deletion failed. Please contact admin" });
+    }
+
```
Comment (Achintya-Chatterjee) [inline controllers/external-accounts.js:85]: We can improve the error messages by providing detailed messages that help debug the issue in production.

File: test/unit/utils/removeDiscordRole.test.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  it("should throw an error if role deletion failed", async function () {
+    discordId = "randomDiscordId";
+
+    try {
+      await removeDiscordRole(userData[0], discordId, roleid, rolename);
+    } catch (error) {
+      expect(error.message).to.equal("Role deletion failed");
+    }
+  });
+});
```
Comment (Achintya-Chatterjee) [inline test/unit/utils/removeDiscordRole.test.js:87]: the tests for error conditions where both `roleid` and `rolename` are undefined and are missing. Adding a test to handle this edge case is important to ensure no unexpected behaviour happens.

File: utils/removeDiscordRole.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    const { wasSuccess } = await discordRolesModel.removeMemberGroup(roleData.roleid, discordId);
+    if (!wasSuccess) {
+      throw new Error("Role deletion failed");
+    }
+
+    return true;
+  } catch (error) {
+    logger.error(`Error getting external account data: ${error}`);
+    return false;
```
Comment (Achintya-Chatterjee) [inline utils/removeDiscordRole.js:31]: In the `catch` block, instead of a generic log statement, we should log more details about the failure, like which role for which the user failed and what parameters were passed. 

```suggestion
  } catch (error) {
    llogger.error(`Error removing role ${rolename || roleid} for user ${discordId}: ${error.message}`);

    return false;
```

**Review by Achintya-Chatterjee — CHANGES_REQUESTED** [review]

File: utils/removeDiscordRole.js [hunk 6]
```diff
@@ -27,7 +27,8 @@ const removeDiscordRole = async (userData, discordId, roleid, rolename) => {
 
     return true;
   } catch (error) {
-    logger.error(`Error getting external account data: ${error}`);
+    llogger.error(`Error removing role ${rolename || roleid} for user ${discordId}: ${error.message}`);
```
Comment (shubhdevelop) [inline utils/removeDiscordRole.js:30]: typo

File: controllers/external-accounts.js (hunk 3 above)
Comment (shobhan-sundar-goutam) [reply controllers/external-accounts.js]: I've added the changes. Can you please check it?

File: test/unit/utils/removeDiscordRole.test.js (hunk 4 above)
Comment (shobhan-sundar-goutam) [reply test/unit/utils/removeDiscordRole.test.js:98]: I've added the changes. Can you please check it?

File: utils/removeDiscordRole.js (hunk 6 above)
Comment (shobhan-sundar-goutam) [reply utils/removeDiscordRole.js]: fixed that

File: constants/discordRoles.ts
Comment (yesyash) [inline constants/discordRoles.ts:1]: - are we not storing other discord roles in backend already? do we need to create a new file?

File: constants/discordRoles.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+const DISCORD_ROLES = {
+  UNVERIFIED: "unverified",
+};
+
+module.exports = {
+  DISCORD_ROLES,
+};
```
Comment (yesyash) [inline constants/discordRoles.ts:7]: - instead of module.exports can we export the object directly?
```js
export const DISCORD_ROLES = {
    ...
}
```

File: utils/removeDiscordRole.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+/*
+  This function handles all scenarios for role arguments - only "roleid" / only "rolename" / both.
+  Ensure that if only "roleid" is provided, pass "rolename" as "undefined", and vice versa.
+  When both "roleid" and "rolename" are available, pass both as arguments.
+  For implementation reference, see linkExternalAccount in controllers/external-accounts.js."
+ */
```
Comment (yesyash) [inline utils/removeDiscordRole.js:6]: - let's make this short and please follow jsdoc conventions.

File: utils/removeDiscordRole.js
Comment (yesyash) [inline utils/removeDiscordRole.js:1]: - do we have a file for all discord related utils? If so can we move it there instead of creating a new file in utils?

File: controllers/external-accounts.js [hunk 9]
```diff
@@ -69,6 +72,19 @@ const linkExternalAccount = async (req, res) => {
...
+    const unverifiedRoleRemovalResponse = await removeDiscordRoleUtils.removeDiscordRole(
+      req.userData,
+      attributes.discordId,
+      undefined,
+      DISCORD_ROLES.UNVERIFIED
+    );
+
+    if (!unverifiedRoleRemovalResponse.success) {
+      return res.status(500).json({
+        message: `User details updated successfully but ${unverifiedRoleRemovalResponse.message}. Please contact admin`,
```
Comment (yesyash) [inline controllers/external-accounts.js:84]: - if the action failed how did is "user details updated successfully" the correct sentence?

File: utils/removeDiscordRole.js
Comment (shobhan-sundar-goutam) [reply utils/removeDiscordRole.js:1]: Tried to add this function within discord-actions util file but getting this problem. It's like Module A is calling Module B and then Module B is calling Module A for some other dependent function but not able to do that. Node doesn't allow module exports inside circular dependency after v14 I guess.
That's why creating a new file seemed a good choice to me.
Please let me know if I'm wrong somewhere or any other way to solve this issue

image

File: constants/discordRoles.ts (hunk 7 above)
Comment (shobhan-sundar-goutam) [reply constants/discordRoles.ts]: Exported the object directly as you asked.

File: utils/removeDiscordRole.js (hunk 8 above)
Comment (shobhan-sundar-goutam) [reply utils/removeDiscordRole.js]: Rewritten it using jsdoc convention. Can you please check if it is correct?

File: controllers/external-accounts.js (hunk 9 above)
Comment (shobhan-sundar-goutam) [reply controllers/external-accounts.js]: Removed the word "successfully". Hope it is correct now. Please let me know if the message makes sense now.

File: utils/removeDiscordRole.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+    return { success: true, message: "Role deleted successfully" };
+  } catch (error) {
+    logger.error(`Error removing role ${rolename || roleid} for user ${discordId}: ${error.message}`);
+
+    return { success: false, message: error.message };
+  }
+};
+
+module.exports = { removeDiscordRole };
```
Comment (yesyash) [inline utils/removeDiscordRole.js:39]: - can we directly `removeDiscordRole` instead of doing module.export

Comment (shobhan-sundar-goutam) [reply utils/removeDiscordRole.js]: Made the changes as you asked.

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by yesyash — APPROVED** [review]
